### PR TITLE
Setup typescript

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**

### CODE REVIEW

**Comment by wadjih-bencheikh18** [comment]
done

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+}
+export interface Criterion {
+  is: boolean | RegExp;
+  key: boolean | RegExp | string;
+  negate: boolean;
+  valueReg: boolean | undefined;
+  checkString: (arg: string) => boolean;
+  checkNumber: (arg: number) => boolean;
+}
+export interface ArrayType {
```
Comment (stropitek) [inline src/index.ts:28]: What is this type for??

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ * @param [options.caseSensitive=false] - By default we ignore case.
+ * @param [options.ignorePaths=[]] - Array of jpath to ignore.
+ * @param [options.pathAlias={}] - Key (string), value (string of regexp).
+ * @param [options.keywords=[]] - List of keywords used to filter the array.
+ * @param [options.index=false] - Returns the indices in the array that match.
+ * @param [options.predicate='AND'] - Could be either AND or OR.
+ * @returns String[] | number[].
+ */
+export function filter(
+  array: ArrayType[] | string[] | number[] | { [s: string | number]: any },
```
Comment (stropitek) [inline src/index.ts:59]: can you give this parameter with a meaningful name? For example just `data` would be better.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  options: OptionsType | undefined = {
+    ignorePaths: [],
+    pathAlias: {},
+  },
```
Comment (stropitek) [inline src/index.ts:63]: You already assign default values when destructuring later
```suggestion
  options: OptionsType = {}},
```

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  let insensitive = options.caseSensitive ? '' : 'i';
+
+  pathAlias = ensureObjectOfRegExps(pathAlias, { insensitive });
+  ignorePaths = ignorePaths.map((path) =>
+    typeof path === 'string'
+      ? new RegExp(`(^|\\.)${escapeRegExp(path)}(\\.|$)`, insensitive)
+      : path,
+  );
+
+  let limit = options.limit ? options.limit : Infinity;
```
Comment (stropitek) [inline src/index.ts:82]: You can move this above in the same statement as for the other default values

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  let keywords = options.keywords || [];
```
Comment (stropitek) [inline src/index.ts:84]: ditto

File: src/match/match.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * @param options.pathAlias - { [s: string]: RegExp|string }.
+ * @returns Boolean.
+ */
+export default function match(
+  element: ArrayType | string | number,
+  criteria: Criterion[],
+  predicate: string,
+  options: {
+    ignorePaths: RegExp[];
+    pathAlias: { [s: string]: RegExp | string };
```
Comment (stropitek) [inline src/match/match.ts:22]:
```suggestion
    pathAlias: Record<string, RegExp | string>;
```

File: src/match/recursiveMatch.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (stropitek) [inline src/match/recursiveMatch.ts:1]: There are no explicit any in this file

File: src/match/recursiveMatch.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * @param element - ArrayType.
+ * @param criterium - Criterion.
+ * @param keys - String[].
+ * @param options - Object.
+ * @param options.ignorePaths - RegExp[].
+ * @param options.pathAlias - { [s: string]: RegExp }.
+ * @returns Boolean.
+ */
+export default function recursiveMatch(
+  element: ArrayType | string | number | any,
```
Comment (stropitek) [inline src/match/recursiveMatch.ts:18]: I think there shouldn't be an `any` here

File: src/utils/ensureObjectOfRegExps.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * EnsureObjectOfRegExps.
+ *
+ * @param object  - { [index: string]: string|RegExp }.
+ * @param options - Object.
+ * @param options.insensitive - String.
+ * @returns - Result.
+ */
+export default function ensureObjectOfRegExps(
+  object: { [index: string]: string | RegExp },
```
Comment (stropitek) [inline src/utils/ensureObjectOfRegExps.ts:12]:
```suggestion
  object: Record<string, string | RegExp>,
```

File: src/utils/ensureObjectOfRegExps.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  options: { insensitive: string },
+) {
+  const { insensitive } = options;
+  const toReturn: { [index: string]: RegExp } = {};
```
Comment (stropitek) [inline src/utils/ensureObjectOfRegExps.ts:16]:
```suggestion
  const toReturn: Record<string, RegExp> = {};
```

File: src/utils/getCheckNumber.ts [hunk 11]
```diff
@@ -1,4 +1,6 @@
-const operators = {
+const operators: {
+  [index: string]: (arg1: number) => (arg: number) => boolean;
+} = {
```
Comment (stropitek) [inline src/utils/getCheckNumber.ts:3]:
```suggestion
const operators: Record<string, (arg1: number) => (arg: number) => boolean> = {
```

File: src/utils/getCheckString.ts [hunk 12]
```diff
@@ -41,12 +43,23 @@ const operators = {
...
+  insensitive: string,
+): (arg: string) => boolean {
   let parts = keyword.split('..');
+  // eslint-disable-next-line prefer-named-capture-group
   let match = /^\s*\(?\s*(<|<=|=|>=|>)?(\S*)\s*\)?$/.exec(parts[0]);
-  let checkString = () => false;
+  let checkString: (arg: string) => boolean = () => false;
   if (match) {
-    let operator = match[1];
+    let operator: string = match[1];
```
Comment (stropitek) [inline src/utils/getCheckString.ts:62]:
```suggestion
    let operator = match[1];
```

File: src/index.ts (hunk 2 above)
Comment (stropitek) [reply src/index.ts:59]: I think we could use typescript's recursion feature to do something better.

```suggestion
  array: ArrayType[] | string[] | number[] | Record<string | number, any>,
```

**Review by stropitek — COMMENTED** [review]
As part of this PR, could you also please replace all `let` statements with `const` and then use `let` only where variables are reassigned?

File: package.json [hunk 13]
```diff
@@ -30,25 +32,31 @@
   "license": "MIT",
   "scripts": {
     "build": "cheminfo-build --entry src/index.js --root SmartArrayFilter",
```
Comment (stropitek) [inline package.json:34]: This won't work (see https://github.com/mljs/global-spectral-deconvolution/blob/e9e8522996bbe435e7778dd910c304cdfc578cae/package.json#L17)

File: src/index.ts (hunk 1 above)
Comment (wadjih-bencheikh18) [reply src/index.ts:28]: https://github.com/cheminfo/smart-array-filter/blob/9e8b69a246556deb7cf769a0ddcd146bf993bbb0/src/__tests__/index.test.js#L5-L30
ArrayType is for this

File: src/index.ts (hunk 3 above)
Comment (wadjih-bencheikh18) [reply src/index.ts:63]: Yes this was an error

File: src/match/recursiveMatch.ts (hunk 8 above)
Comment (wadjih-bencheikh18) [reply src/match/recursiveMatch.ts:18]: Yes i'm going to use typescript recursion

Comment (wadjih-bencheikh18) [reply src/match/recursiveMatch.ts:18]: fixed

File: src/match/match.ts (hunk 6 above)
Comment (wadjih-bencheikh18) [reply src/match/match.ts:22]: fixed

File: src/index.ts (hunk 1 above)
Comment (wadjih-bencheikh18) [reply src/index.ts:28]: Deleted

File: src/match/recursiveMatch.ts (hunk 8 above)
Comment (wadjih-bencheikh18) [reply src/match/recursiveMatch.ts]: fixed can you check please the new type

File: src/index.ts (hunk 1 above)
Comment (wadjih-bencheikh18) [reply src/index.ts]: Deleted

File: package.json (hunk 13 above)
Comment (wadjih-bencheikh18) [reply package.json]: fixed can you check this line too please

**Comment by stropitek** [comment]
@wadjih-bencheikh18 have a look at https://github.com/cheminfo/smart-array-filter/pull/23/commits/d5a3ccc29f578f15417e1590f04d7bbfe9a9f36d

- Used a recursive type to avoid using `any`
- Made the main exported function a generic (because the type of the output array can be the same as the type of the input array)
- Made an overload so that when the option `index` is set to `true`, the returned type is an array of numbers

**Pull request merged**